### PR TITLE
Ledger notifications refactor and add new BTC=>BAT notification

### PR DESCRIPTION
**Pull request by bsclifton**
screen shot 2017-10-03 at 4 48 39 pm

Fixes https://github.com/brave/browser-laptop/issues/11021

Presents this one-time "Converted BTC to BAT" alert:
 - if payments are enabled
 - user has a positive balance
 - this is an existing profile (new profiles will have firstRunTimestamp matching btcToBatTimestamp)
 - notification has not already been shown yet

Includes a "learn more" link pointing to:
https://brave.com/faq-payments/#brave-payments

Has a possible security risk- the deep link allows URL hash to modify state (limited to boolean state variables on about:preferences)

Auditors: @NejcZdovc, @diracdeltas

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).

## Automated Test Plan:
`npm run unittest -- --grep="ledger api unit tests"`
`npm run unittest -- --grep="ledgerReducer unit tests"`
`npm run unittest -- --grep="Preferences component"`

## Reviewer Checklist:
- [x] Adequate test coverage exists to prevent regressions
- [x] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [x] New files have MPL2 license header

## Merge Checklist:
- [x] Merge to master
- [ ] rebase to 0.20.x (taking care to make sure setting names match)
- [ ] cherry-pick the 0.20.x rebase to 0.19.x
- [ ] list commit hashes here

### CODE REVIEW

File: js/about/preferences.js [hunk 1]
```diff
@@ -759,13 +766,42 @@ class AboutPreferences extends React.Component {
...
+      newState[params] = true
+    }
+    this.setState(newState)
   }
 
+  /**
+   * Parses a query string like:
+   * about:preferences#payments?ledgerBackupOverlayVisible
+   * and returns the part:
+   * `about:preferences#payments`
```
Comment (diracdeltas) [inline js/about/preferences.js]: This comment doesn't seem right. Previously this would return 'payments' not 'about:preferences#payments'.

Comment (bsclifton) [reply js/about/preferences.js]: Good catch; fixed 😄

File: app/browser/api/ledger.js [hunk 2]
```diff
@@ -104,13 +102,264 @@ const fileTypes = {
...
+    }
+  },
+  onInterval: (state) => {
+    if (!getSetting(settings.PAYMENTS_NOTIFICATIONS)) {
+      return
+    }
+    if (getSetting(settings.PAYMENTS_ENABLED)) {
+      notifications.showEnabledNotifications(state)
+    } else {
+      notifications.showDisabledNotifications(state)
```
Comment (bbondy) [inline app/browser/api/ledger.js]: before we showed this notification only when `!getSetting(settings.PAYMENTS_NOTIFICATIONS)` is that change intentional?

Comment (bsclifton) [reply app/browser/api/ledger.js]: logic shouldn't have changed at all (just moved). If PAYMENT_NOTIFICATIONS is set, none of the notifications should be shown

File: app/browser/api/ledger.js [hunk 3]
```diff
@@ -104,13 +102,264 @@ const fileTypes = {
...
+            windowId: activeWindow.id
+          })
+        }
+        break
+
+      default:
+        return
+    }
+
+    appActions.hideNotification(message)
```
Comment (bbondy) [inline app/browser/api/ledger.js]: maybe the above default: should be just break so this always runs?

File: app/browser/api/ledger.js (hunk 2 above)
Comment (bbondy) [reply app/browser/api/ledger.js]: I guess I'm thinking about getSetting(settings.PAYMENTS_NOTIFICATIONS) is false but getSetting(settings.PAYMENTS_ENABLED is true, before it would show.

File: app/browser/api/ledger.js (hunk 3 above)
Comment (bsclifton) [reply app/browser/api/ledger.js]: I considered that... but since other messages may be coming through (since this fires on `ipc.on(messages.NOTIFICATION_RESPONSE`) I didn't want to change the behavior

File: app/browser/api/ledger.js (hunk 2 above)
Comment (bsclifton) [reply app/browser/api/ledger.js]: Great catch- I rolled back to the original behavior (so that there is no change from previous code)

**Comment by bradleyrichter** [comment]
@bsclifton This alert bar location looks wrong. The screen shot above shows the per-tab location and this should be at the app level.

**Comment by luixxiul** [comment]
https://github.com/brave/browser-laptop/issues/11256

**Comment by luixxiul** [comment]
also: https://github.com/brave/browser-laptop/issues/8934

**Comment by bsclifton** [comment]
master https://github.com/brave/browser-laptop/commit/9285507f67b07a2537b2e1581c4c30e3bf2b3748
0.20.x https://github.com/brave/browser-laptop/commit/8f2fc5cefbd9d8c3c24391d66bfeb64c6e471440
0.19.x https://github.com/brave/browser-laptop/commit/e84899a1390d4c5e2bc4def1e09272105a5e3c62
